README.md
# flutter_web_struct
Package for help developper to make a nice website!

## Installation
Add `flutter_web_struct` as a dependency in your pubspec.yaml file.

```yaml
dependencies:
flutter_web_struct: ^0.0.2
```

## Usage
For use flutter_web_struct by following steps : 

```dart
// Import the plugin
import 'package:flutter_web_struct/flutter_web_struct.dart';

// call initialisation in main 
void main() async {
  runApp(const MyApp());
} 
```

Call WebStructController and configure your Design


```dart
class MyApp extends StatelessWidget {
  MyApp({Key? key}) : super(key: key);
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
   
    var ctrl = WebStructController();
    //Configure Base Url if you need
    ctrl.webStructUrlBase = "http://www.mywebsite.com";

    // this is header
    ctrl.header = Container(
      color: Colors.red,
      height: 48,
      child: const Center(
        child: Text("This is header"),
      ),
    );
 // this is footer
    ctrl.footer = Container(
      color: Colors.white,
      height: 24,
      child: const Center(
        child: Text(
          "This is footer",
        ),
      ),
    );
    
    //you can manage your MaterialApp as you want, with your RouteManager

    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage('Flutter Demo Home Page'),
    );
  }
}

```

Create your page  extends WebPage
```dart

class MyHomePage extends WebPage {
// Choose FULLCONTENT or SCROLLVIEW (defaut ScrollView)
  MyHomePage(title) : super(title, typePage: TypePage.FULLCONTENT);

//Complet your design for Large screen
  @override
  Widget largeContent(BuildContext context) {
    return Container(
      color: Colors.blue,
      child: const Center(
        child: Text(
          "Large content",
        ),
      ),
    );
  }

//Complet your design for Small screen

  @override
  Widget smallContent(BuildContext context) {
    return Container(
      color: Colors.green,
      child: const Center(
        child: Text(
          "Small content",
        ),
      ),
    );
  }

//Complet your design for Medium screen
  @override
  Widget mediumContent(BuildContext context) {
    return Container(
      color: Colors.purple,
      child: const Center(
        child: Text(
          "Medium content",
        ),
      ),
    );
  }
}

```


Or Create your page  extends MobilePage
```dart


class MyMobilePage extends MobilePage{
  MyMobilePage(String title) : super(title,typePage: TypePage.FULLCONTENT,typeMenu: TypeMenu.DRAWER);

  @override
  Widget largeContent(BuildContext context) {
    return  Container(
      color: Colors.blue,
      child: const Center(
        child: Text(
          "Large & medium content",
        ),
      ),
    );
  }

  @override
  Widget smallContent(BuildContext context) {
    return  Container(
      color: Colors.blue,
      child: const Center(
        child: Text(
          "Small content",
        ),
      ),
    );
  }
}

```

Configure your screen 
```dart
    ctrl.webStructLargePage = 1200; //default value
    ctrl.webStructMediumPage = 800; //default value
```

CHANGELOG.md
## 0.0.3
//TODO



## 0.0.2
 add Structure only Large and Small


## 0.0.1
First developpement


lib/src/webpage.dart
import 'package:flemme/flemme.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_web_struct/src/responsive_widget.dart';

import 'material_web_app.dart';

abstract class WebPage extends StatelessWidget {
  WebPage(this.title, {this.typePage = TypePage.SCROLLVIEW}) : super();

  late String title;
  late TypePage typePage;

  @override
  Widget build(BuildContext context) {
    return ResponsiveWidget(
      largeScreen: largeScreen(context),
      smallScreen: smallScreen(context),
      mediumScreen: mediumScreen(context),
    );
  }

  Scaffold largeScreen(BuildContext context) {
    return Scaffold(
      body: [
        webStructController.header ?? SizedBox(),
        typePage == TypePage.SCROLLVIEW
            ? SingleChildScrollView(
                child: [
                  largeContent(context),
                  webStructController.footer ?? const SizedBox()
                ].listToColumn(),
              ).withExpanded()
            : [
                largeContent(context).withFlexible(),
                webStructController.footer ?? const SizedBox()
              ].listToColumn().withExpanded(),
      ].listToColumn(),
    );
  }

  Scaffold mediumScreen(BuildContext context) {
    return Scaffold(
      body: [
        webStructController.header ?? SizedBox(),
        typePage == TypePage.SCROLLVIEW
            ? SingleChildScrollView(
                child: [
                  mediumContent(context),
                  webStructController.footer ?? const SizedBox()
                ].listToColumn(),
              ).withExpanded()
            : [
                mediumContent(context).withFlexible(),
                webStructController.footer ?? const SizedBox()
              ].listToColumn().withExpanded(),
      ].listToColumn(),
    );
  }

  Scaffold smallScreen(BuildContext context) {
    return Scaffold(
      appBar: webStructController.appBar ??
          AppBar(
            title: Text(title),
          ),
      drawer: webStructController.drawer,
      body: [
        typePage == TypePage.SCROLLVIEW
            ? SingleChildScrollView(
                child: [
                  smallContent(context),
                  webStructController.footer ?? const SizedBox()
                ].listToColumn(),
              ).withExpanded()
            : [
                smallContent(context).withFlexible(),
                webStructController.footer ?? const SizedBox()
              ].listToColumn().withExpanded(),
      ].listToColumn(),
    );
  }

  Widget largeContent(BuildContext context);
  Widget mediumContent(BuildContext context);
  Widget smallContent(BuildContext context);
}


lib/src/mobilepage.dart
import 'package:flemme/flemme.dart';
import 'package:flutter/material.dart';

import '../flutter_web_struct.dart';

abstract class MobilePage extends StatelessWidget {
  const MobilePage(this.title,
      {this.typePage = TypePage.SCROLLVIEW,
      this.typeMenu = TypeMenu.DRAWER,
      Key? key})
      : super(key: key);

  final String title;
  final TypePage typePage;
  final TypeMenu typeMenu;

  @override
  Widget build(BuildContext context) {
    return ResponsiveWidget(
      largeScreen: largeScreen(context),
      smallScreen: smallScreen(context),
      mediumScreen: largeScreen(context),
    );
  }

  Scaffold largeScreen(BuildContext context) {
    return Scaffold(
      drawer: typeMenu == TypeMenu.DRAWER ? webStructController.drawer : null,
      body: [
        typeMenu == TypeMenu.HEADER
            ? (webStructController.header ?? const SizedBox())
            : const SizedBox(),
        typePage == TypePage.SCROLLVIEW
            ? SingleChildScrollView(
                child: [
                  largeContent(context),
                  webStructController.footer ?? const SizedBox()
                ].listToColumn(),
              ).withExpanded()
            : [
                largeContent(context).withFlexible(),
                webStructController.footer ?? const SizedBox()
              ].listToColumn().withExpanded(),
      ].listToColumn(),
    );
  }

  Scaffold smallScreen(BuildContext context) {
    return Scaffold(
      appBar: webStructController.appBar ??
          AppBar(
            title: Text(title),
          ),
      drawer: webStructController.drawer,
      body: [
        typePage == TypePage.SCROLLVIEW
            ? SingleChildScrollView(
                child: [
                  smallContent(context),
                  webStructController.footer ?? const SizedBox()
                ].listToColumn(),
              ).withExpanded()
            : [
                smallContent(context).withFlexible(),
                webStructController.footer ?? const SizedBox()
              ].listToColumn().withExpanded(),
      ].listToColumn(),
    );
  }

  Widget largeContent(BuildContext context);
  Widget smallContent(BuildContext context);
}


lib/src/responsive_widget.dart
import 'package:flutter/material.dart';
import 'package:flutter_web_struct/flutter_web_struct.dart';

// ignore: must_be_immutable
class ResponsiveWidget extends StatefulWidget {
  late Widget largeScreen;
  Widget? mediumScreen;
  Widget? smallScreen;

  ResponsiveWidget(
      {required this.largeScreen,
      this.mediumScreen,
      this.smallScreen,
      Key? key})
      : super(key: key);

  @override
  _ResponsiveWidgetState createState() => _ResponsiveWidgetState();

  static bool isLargeScreen(BuildContext context) {
    return MediaQuery.of(context).size.width >
        webStructController.webStructLargePage;
  }

  //Small screen is any screen whose width is less than 800 pixels
  static bool isSmallScreen(BuildContext context) {
    return MediaQuery.of(context).size.width <
        webStructController.webStructMediumPage;
  }

  //Medium screen is any screen whose width is less than 1200 pixels,
  //and more than 800 pixels
  static bool isMediumScreen(BuildContext context) {
    return MediaQuery.of(context).size.width >
            webStructController.webStructMediumPage &&
        MediaQuery.of(context).size.width <
            webStructController.webStructLargePage;
  }
}

class _ResponsiveWidgetState extends State<ResponsiveWidget> {
  @override
  Widget build(BuildContext context) {
    //Returns the widget which is more appropriate for the screen size
    return LayoutBuilder(builder: (context, constraints) {
      if (constraints.maxWidth > webStructController.webStructLargePage) {
        return widget.largeScreen;
      } else if (constraints.maxWidth >
              webStructController.webStructMediumPage &&
          constraints.maxWidth < webStructController.webStructLargePage) {
        //if medium screen not available, then return large screen
        return widget.mediumScreen ?? widget.largeScreen;
      } else {
        //if small screen implementation not available, then return large screen
        return widget.smallScreen ?? widget.largeScreen;
      }
    });
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_web_struct/flutter_web_struct.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  MyApp({Key? key}) : super(key: key);
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    var ctrl = WebStructController();

    ctrl.webStructUrlBase = "http://www.mywebsite.com";
    ctrl.header = Container(
      color: Colors.red,
      height: 48,
      child: const Center(
        child: Text("This is header"),
      ),
    );
    ctrl.footer = Container(
      color: Colors.white,
      height: 24,
      child: const Center(
        child: Text(
          "This is footer",
        ),
      ),
    );

    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage('Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends WebPage {
  MyHomePage(title) : super(title, typePage: TypePage.FULLCONTENT);

  @override
  Widget largeContent(BuildContext context) {
    return Container(
      color: Colors.blue,
      child: const Center(
        child: Text(
          "Large content",
        ),
      ),
    );
  }

  @override
  Widget smallContent(BuildContext context) {
    return Container(
      color: Colors.green,
      child: const Center(
        child: Text(
          "Small content",
        ),
      ),
    );
  }

  @override
  Widget mediumContent(BuildContext context) {
    return Container(
      color: Colors.purple,
      child: const Center(
        child: Text(
          "Medium content",
        ),
      ),
    );
  }
}

class MyMobilePage extends MobilePage {
  MyMobilePage(String title)
      : super(title, typePage: TypePage.FULLCONTENT, typeMenu: TypeMenu.DRAWER);

  @override
  Widget largeContent(BuildContext context) {
    return Container(
      color: Colors.blue,
      child: const Center(
        child: Text(
          "Large & medium content",
        ),
      ),
    );
  }

  @override
  Widget smallContent(BuildContext context) {
    return Container(
      color: Colors.blue,
      child: const Center(
        child: Text(
          "Small content",
        ),
      ),
    );
  }
}
